File: java/epi/Anagrams.java
package epi;
import test_framework.EpiTest;
import test_framework.EpiTestComparator;
import test_framework.LexicographicalListComparator;
import test_framework.GenericTest;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BiPredicate;

public class Anagrams {

  @EpiTest(testDataFile = "../test_data/epi/anagrams.tsv")
  public static List<List<String>> findAnagrams(List<String> dictionary) {
    List<List<String>> ans = new ArrayList<>();
    HashMap<String, List<String>> m = new HashMap<>();

    for (String word : dictionary) {
      char[] sArr = word.toCharArray();
      Arrays.sort(sArr);
      String sWord = new String(sArr);
      m.putIfAbsent(sWord, new ArrayList<>());
      m.get(sWord).add(word);
    }

    for (Map.Entry<String, List<String>> e : m.entrySet()) {
      if (e.getValue().size() >= 2) {
        ans.add(e.getValue());
      }
    }
  
    return ans;
  }

  @EpiTestComparator
  public static BiPredicate<List<List<String>>, List<List<String>>> comp =
      (expected, result) -> {
    if (result == null) {
      return false;
    }
    for (List<String> l : expected) {
      Collections.sort(l);
    }
    expected.sort(new LexicographicalListComparator<>());
    for (List<String> l : result) {
      Collections.sort(l);
    }
    result.sort(new LexicographicalListComparator<>());
    return expected.equals(result);
  };

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "Anagrams.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: cpp/epi/anagrams.cc
#include <algorithm>
#include <functional>
#include <string>
#include <unordered_map>
#include <vector>

#include "generic_test.h"

using std::hash;
using std::sort;
using std::string;
using std::to_string;
using std::unordered_map;
using std::vector;

vector<vector<string>> FindAnagrams(const vector<string>& dictionary) {
  return {};
}


int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::string func_name = "FindAnagrams";
  std::vector<std::string> param_names{"dictionary"};

  return GenericTestMain(
      args, "anagrams.cc", "../test_data/epi/anagrams.tsv", &FindAnagrams,
      &UnorderedComparator<std::vector<std::vector<std::string>>>, func_name, param_names);
}
